### DISCUSSION TITLE
How to "mock" custom Font-Resources in headless testing

### DISCUSSION THREAD

**Discussion by JanZsch** [AvaloniaUI/Avalonia#17451] [Q&A]
Hi.
In my avalonia app I'm using font awesome 6 to depict icons. I integrated the font as is described in the docs and in this [github post](https://github.com/AvaloniaUI/Avalonia/discussions/16736) using an avalonia resource. 

Now, I want to test my app E2E using the headless mode in conjunction with Reqnroll/SpecFlow, i.e. in the test setup I configure my app by:
```cs
this.appBuilder = AppBuilder
        .Configure<App>()
        .UseHeadless(new AvaloniaHeadlessPlatformOptions() { UseHeadlessDrawing = true })
        .StartWithClassicDesktopLifetime();

((IClassicDesktopStyleApplicationLifetime)appBuilder.Instance.ApplicationLifetime).MainWindow.Show();
```
The moment the Show() method is called, I get an invalid operation exception, basically stating that my font awesome font could not be created (to be precise: 'Could not create glyphTypeface. Font family: Font awesome 6 ... (key: avares://...)'). What I found from debugging is that in headless mode the 'HeadlessFontManagerStub' is used for the 'IFontManagerImpl' interface. This stub registeres non-system fonts using HeadlessGlyphTypefaceImpl with family name "$Default" using the method TryCreateGlyphTypeface(Stream ..., FontSimulation). Not using the correct font family name here but "$Default" then leads to the exception as the actually queried font key, namely Font Awesome 6, is not registered with the FontManager.

Now the question: Is there a way to allow custom fonts in your app and test it headlessly the same time? My google search did not bring up anything, which made me wonder if my approach is maybe totally of. Also note, that as I want to use SpecFlow/Reqnroll I cannot use the test attribute `[AvaloniaTest]`.

Any help or suggestions would be very much appreciated.

**Comment by stevemonaco** [comment]
> Also note, that as I want to use SpecFlow/Reqnroll I cannot use the test attribute [AvaloniaTest].

This part seems problematic. The Avalonia-related headless test attributes exist to configure the test runner so that tests:
1. use the same, shared Headless session.
2. run via the `Dispatcher` on the UI thread.
3. run one at a time (no parallel).
4. can resume awaits on the UI thread.
5. are cleaned up with `Dispatcher.UI.RunJobs`.
